tests/integration_tests/conftest.py

```diff
@@ -80,8 +81,61 @@ def disable_subp_usage(request):
     pass
 
 
+def setup_image_or_die(cloud: IntegrationCloud):
+    try:
+        setup_image(cloud)
+    except Exception as e:
+        if cloud.snapshot_id:
+            # if a snapshot id was set, then snapshot succeeded, teardown
+            cloud.delete_snapshot()
+        cloud.destroy()
+        pytest.exit(
+            f"{type(e).__name__} in session setup: {str(e)}", returncode=2
+        )
+
+
+def setup_image_once(
+    testrun_uid: str,
+    cloud: IntegrationCloud,
+    tmp_path_factory: pytest.TempPathFactory,
+) -> None:
+    """Setup image to be used for (almost) all tests.
+
+    Since pytest-xdist runs tests in parallel, we need to ensure that
+    the image setup is only done once, and not multiple times in parallel.
+    """
+    if testrun_uid == "master":
+        # We're running single-threaded, so no synchronization needed
+        setup_image_or_die(cloud)
+        return
+
+    # We're an xdist worker, so we need to synchronize.
+    # Whoever gets the lock first will do the setup, writing the
+    # image id into the image path. The other workers will
+    # wait for file access, read the image id from image path and use it.
+    image_path = Path(
+        tmp_path_factory.getbasetemp().parent,
+        f"session_image_id_{testrun_uid}",
+    )
+    lock_path = image_path.with_suffix(".lock")
+
+    with open(lock_path, "w") as lock_file:
+        # Use a lock to ensure only one worker does the setup
+        fcntl.lockf(lock_file, fcntl.LOCK_EX)
+        try:
+            if not image_path.exists():
+                setup_image_or_die(cloud)
+                image_path.write_text(cloud.image_id)
+            else:
+                cloud.snapshot_id = image_path.read_text().strip()
+        finally:
+            fcntl.lockf(lock_file, fcntl.LOCK_UN)
+
+
 @pytest.fixture(scope="session")
-def session_cloud(reaper: Reaper) -> Generator[IntegrationCloud, None, None]:
+def session_cloud(
+    testrun_uid, reaper: Reaper, tmp_path_factory
```

Comment:
Per the [xdist docs you linked me](https://pytest-xdist.readthedocs.io/en/stable/how-to.html), this actually needs to be the `worker_id` instead of `testrun_uid` (which is just a randomly generated uid instead of a named worker (gw0, gw1 or master). In testing I confirmed that testrun_uid in fact was a UID in all cases both with `-n auto` and without specifying worker threads with `-n`.

Without this change, testrun_uid will never match "master" as it's a UID.
```suggestion
    reaper: Reaper, tmp_path_factory, worker_id
```